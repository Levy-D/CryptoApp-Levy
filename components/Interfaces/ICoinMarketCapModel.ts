/* eslint-disable semi */
export interface Status {
	timestamp: string;
	error_code: number;
	error_message: string;
	elapsed: number;
	credit_count: number;
	notice: string;
	total_count: number;
}

export interface CmcCryptoCurrency {
	id: number;
	name: string;
	symbol: string;
	slug: string;
	num_market_pairs: number;
	date_added: string;
	tags: string[];
	max_supply: number | null;
	circulating_supply: number;
	total_supply: number;
	platform: Platform | null;
	cmc_rank: number;
	last_updated: string;
	quote: Quote;
	isFavorite?: boolean;
}

export interface Platform {
	id: number;
	name: string;
	symbol: string;
	slug: string;
	token_address: string;
}

export interface Quote {
	USD?: Valuta;
	EUR?: Valuta;
}

export interface Valuta {
	price: number;
	volume_24h: number;
	volume_change_24h: number;
	percent_change_1h: number;
	percent_change_24h: number;
	percent_change_7d: number;
	percent_change_30d: number;
	percent_change_60d: number;
	percent_change_90d: number;
	market_cap: number;
	market_cap_dominance: number;
	fully_diluted_market_cap: number;
	last_updated: string;
}

export default interface CMCResponse {
	status: Status;
	data: CmcCryptoCurrency[];
}
