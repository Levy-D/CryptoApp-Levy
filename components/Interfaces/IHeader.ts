/* eslint-disable semi */
export default interface Props {
    title: string
}
