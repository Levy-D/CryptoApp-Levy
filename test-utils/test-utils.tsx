import React from 'react';
import {render as rtlRender} from '@testing-library/react-native';
import {combineReducers, configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';
// Import your own reducer
import userSettingsReducer from '../components/Redux/Slices/UserSettings';
import cryptoDataReducer from '../components/Redux/Slices/CryptoData';

const reducers = combineReducers({
	userSettings: userSettingsReducer,
	cryptoData: cryptoDataReducer,
});

function render(
	ui: any,
	{
		preloadedState,
		store = configureStore({reducer: reducers, preloadedState}),
		...renderOptions
	}: any = {},
) {
	function Wrapper({children}: any) {
		return <Provider store={store}>{children}</Provider>;
	}

	return rtlRender(ui, {wrapper: Wrapper, ...renderOptions});
}

// Re-export everything
export * from '@testing-library/react-native';
// Override render method
export {render};
