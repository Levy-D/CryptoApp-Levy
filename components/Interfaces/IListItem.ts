/* eslint-disable semi */
import {CmcCryptoCurrency} from './ICoinMarketCapModel'

export default interface Props {
    item?: any
    onFavorite?: any
    data?: CmcCryptoCurrency[]
}
