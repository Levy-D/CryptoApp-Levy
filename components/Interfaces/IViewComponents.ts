import {CmcCryptoCurrency} from './ICoinMarketCapModel';

export interface IHeader {
    title: string
}

export interface IListItem {
    item: CmcCryptoCurrency
    navigation:
}

export interface IFlatList {
    data: CmcCryptoCurrency[],
    navigation:
}
