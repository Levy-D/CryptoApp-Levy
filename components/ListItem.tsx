import React, { useState } from 'react';
import { StyleSheet, View, Text, TouchableOpacity } from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import IListItem from './Interfaces/IListItem';
import AsyncStorage from '@react-native-async-storage/async-storage';

const ListItem = ({ item, onFavorite }: IListItem) => {

    let price = item.quote.USD.price;
    price = parseFloat(price).toFixed(2);
    let percent_24h = item.quote.USD.percent_change_24h;
    percent_24h = parseFloat(percent_24h).toFixed(2);

    const [favoriteValue, setFavorite] = useState(item.id);

    const saveValue = async (storageId: string) => {
        try {
            const jsonValue = JSON.stringify(favoriteValue);
            await AsyncStorage.setItem(storageId, jsonValue);
            console.log(storageId, "added");
            console.log("All Keys", await AsyncStorage.getAllKeys());
        } catch (e) {
            console.log(e);
        }
    }

    const deleteValue = async (storageId: string) => {
        try {
            await AsyncStorage.removeItem(storageId)
            console.log(storageId, "deleted");
            console.log("Remaining Keys", await AsyncStorage.getAllKeys());
        } catch (e) {
            console.log(e);
        }
    }

    const btnPress = () => {
        try {
            let storageId: string = "ID_" + item.id;
            if (!item.isFavorite) {
                setFavorite(item.id);
                saveValue(storageId);
                onFavorite();
                item.isFavorite = true;
            }
            else {
                deleteValue(storageId);
                onFavorite();
                item.isFavorite = false;
            }
        } catch (e) {
            console.log(e);
        }
    }

    return (
        <TouchableOpacity style={styles.listItem}>
            <View style={styles.listItemView} >
                <Text style={styles.listItemName}>{item.name}</Text>
                <Text style={styles.listItemPrice}>${price}</Text>
                <Text style={(percent_24h < 0) ? styles.listItemPercentNegative : styles.listItemPercentPositive}>{percent_24h}%</Text>
                <Icon style={item.isFavorite ? styles.favIconIsFavoriteTrue : styles.favIconIsFavoriteFalse} name="star" onPress={() => btnPress()} ></Icon>
            </View>
        </TouchableOpacity>
    )
};

const styles = StyleSheet.create({
    listItem: {
        justifyContent: 'center',
        height: 70,
        padding: 15,
        backgroundColor: '#f8f8f8',
        borderBottomWidth: 1,
        borderColor: '#eee'
    },
    listItemView: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center'
    },
    listItemName: {
        fontSize: 20,
        width: 130
    },
    listItemPrice: {
        fontSize: 20,
        width: 100
    },
    listItemPercentNegative: {
        fontSize: 20,
        width: 75,
        textAlign: 'right',
        color: 'red'
    },
    listItemPercentPositive: {
        fontSize: 20,
        width: 75,
        textAlign: 'right',
        color: 'green'
    },
    listItemFavorite: {
        fontSize: 20,
        width: 50,
    },
    favIconIsFavoriteTrue: {
        fontSize: 30,
        color: 'gold'
    },
    favIconIsFavoriteFalse: {
        fontSize: 30,
        color: '#d3d3d3'
    }
});

export default ListItem;