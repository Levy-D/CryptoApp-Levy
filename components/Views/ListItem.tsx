import React, {useState} from 'react';
import {StyleSheet, View, Text, TouchableOpacity} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import IListItem from '../Interfaces/IListItem';
import AsyncStorage from '@react-native-async-storage/async-storage';

const ListItem = ({item, onFavorite}: IListItem) => {
	const price : number = parseFloat(item.quote.USD.price);
	const percent24h : number = parseFloat(item.quote.USD.percent_change_24h);

	const [favoriteValue, setFavorite] = useState<Number>(item.id);

	const saveValue = async (storageId: string) => {
		try {
			const jsonValue : string = JSON.stringify(favoriteValue);
			await AsyncStorage.setItem(storageId, jsonValue);
			console.log(storageId, 'added');
			console.log('All Keys', await AsyncStorage.getAllKeys());
		} catch (e) {
			console.log(e);
		}
	};

	const deleteValue = async (storageId: string) => {
		try {
			await AsyncStorage.removeItem(storageId);
			console.log(storageId, 'deleted');
			console.log('Remaining Keys', await AsyncStorage.getAllKeys());
		} catch (e) {
			console.log(e);
		}
	};

	const btnPress = () => {
		try {
			const storageId: string = 'ID_' + item.id;
			if (!item.isFavorite) {
				setFavorite(item.id);
				saveValue(storageId);
				onFavorite();
				item.isFavorite = true;
			} else if (item.isFavorite) {
				deleteValue(storageId);
				onFavorite();
				item.isFavorite = false;
			}
		} catch (e) {
			console.log(e);
		}
	};

	return (

		<View style={styles.listItemView} >
			<Text style={styles.listItemName}>{item.name}</Text>
			<Text style={styles.listItemPrice}>${price.toFixed(2)}</Text>
			<Text style={(percent24h < 0) ? styles.listItemPercentNegative : styles.listItemPercentPositive}>{percent24h.toFixed(2)}%</Text>
			<TouchableOpacity >
				<Icon style={item.isFavorite ? styles.favIconIsFavoriteTrue : styles.favIconIsFavoriteFalse} name="star" onPress={() => btnPress()} ></Icon>
			</TouchableOpacity>
		</View>

	);
};

const styles = StyleSheet.create({
	listItem: {
		justifyContent: 'center',
		height: 70,
		padding: 15,
		backgroundColor: '#f8f8f8',
		borderBottomWidth: 1,
		borderColor: '#eee',
	},
	listItemView: {
		height: 70,
		padding: 15,
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
	},
	listItemName: {
		fontSize: 20,
		width: 130,
	},
	listItemPrice: {
		fontSize: 20,
		width: 100,
	},
	listItemPercentNegative: {
		fontSize: 20,
		width: 75,
		textAlign: 'right',
		color: 'red',
	},
	listItemPercentPositive: {
		fontSize: 20,
		width: 75,
		textAlign: 'right',
		color: 'green',
	},
	listItemFavorite: {
		fontSize: 20,
		width: 50,
	},
	favIconIsFavoriteTrue: {
		fontSize: 30,
		color: 'gold',
	},
	favIconIsFavoriteFalse: {
		fontSize: 30,
		color: '#d3d3d3',
	},
});

export default ListItem;
