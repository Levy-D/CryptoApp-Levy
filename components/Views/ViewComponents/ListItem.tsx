import React from 'react';
import {StyleSheet, View, Text, TouchableOpacity} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import {IListItem} from '../../Interfaces/IViewComponents';
import {useDispatch, useSelector} from 'react-redux';
import {
	selectStateDisplayEUR,
	selectStateHighlightChainlink,
} from '../../Redux/Slices/UserSettings';
import {
	setCryptoDataItem,
	setIsFavoriteToFalse,
	setIsFavoriteToTrue,
} from '../../Redux/Slices/CryptoData';
import {CmcCryptoCurrency} from '../../Interfaces/ICoinMarketCapModel';

const ListItem = ({item, navigation}: IListItem) => {
	const dispatch = useDispatch();
	const stateDisplayEUR = useSelector(selectStateDisplayEUR);
	const stateHighlightChainlink = useSelector(selectStateHighlightChainlink);
	const valuta = stateDisplayEUR ? 'EUR' : 'USD';
	const valutaSymbol: string = stateDisplayEUR ? 'â‚¬' : '$';
	const {price} = item.quote[valuta]!;
	const percent24h = item.quote[valuta]!.percent_change_24h;

	// Toggle favorites
	const toggleFavorite = (item: CmcCryptoCurrency): void => {
		console.log(item.isFavorite);
		if (!item.isFavorite) {
			dispatch(setIsFavoriteToTrue(item));
		}

		if (item.isFavorite) {
			dispatch(setIsFavoriteToFalse(item));
		}
	};

	const isHighlighted: boolean = item.id === 1975 && stateHighlightChainlink;
	const negativePercent = isHighlighted
		? styles.listItemPercentNegativeChainlink
		: styles.listItemPercentNegative;
	const positivePercent = isHighlighted
		? styles.listItemPercentPositiveChainlink
		: styles.listItemPercentPositive;

	return (
		<TouchableOpacity
			testID="touchable-opacity"
			style={isHighlighted ? styles.listItemChainlink : styles.listItem}
			onPress={() => {
				dispatch(setCryptoDataItem(item));
				navigation.navigate('Crypto Details');
			}}>
			<View style={styles.listItemView}>
				<Text
					testID="name"
					style={
						isHighlighted ? styles.listItemNameChainlink : styles.listItemName
					}>
					{item.name}
				</Text>
				<Text
					testID="price"
					style={
						isHighlighted ? styles.listItemPriceChainlink : styles.listItemPrice
					}>
					{}
					{price < 1
						? `${valutaSymbol}${price.toPrecision(4)}`
						: `${valutaSymbol}${price.toFixed(2)}`}
				</Text>
				<Text
					testID="percent"
					style={percent24h < 0 ? negativePercent : positivePercent}>
					{percent24h.toFixed(2)}%
				</Text>

				<Icon
					testID="icon-star"
					style={
						item.isFavorite
							? styles.favIconIsFavoriteTrue
							: styles.favIconIsFavoriteFalse
					}
					name="star"
					onPress={() => toggleFavorite(item)}></Icon>
			</View>
		</TouchableOpacity>
	);
};

export const styles = StyleSheet.create({
	listItem: {
		justifyContent: 'center',
		fontSize: 16,
		padding: 10,
		backgroundColor: '#f8f8f8',
		borderBottomWidth: 1,
		borderColor: '#eee',
	},
	listItemChainlink: {
		justifyContent: 'center',
		fontSize: 16,
		padding: 10,
		backgroundColor: '#375bd2', // Chainlink blue
		borderBottomWidth: 1,
		borderColor: '#eee',
	},
	listItemNameChainlink: {
		width: 130,
		color: 'white',
	},
	listItemPriceChainlink: {
		width: 100,
		color: 'white',
	},
	listItemPercentNegativeChainlink: {
		width: 75,
		textAlign: 'right',
		color: '#6e0511', // Red
	},
	listItemPercentPositiveChainlink: {
		width: 75,
		textAlign: 'right',
		color: '#8acc26', // Green
	},
	listItemView: {
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
	},
	listItemName: {
		width: 130,
	},
	listItemPrice: {
		width: 100,
	},
	listItemPercentNegative: {
		width: 75,
		textAlign: 'right',
		color: 'red',
	},
	listItemPercentPositive: {
		width: 75,
		textAlign: 'right',
		color: 'green',
	},
	listItemFavorite: {
		width: 50,
	},
	favIconIsFavoriteTrue: {
		fontSize: 24,
		color: 'gold',
	},
	favIconIsFavoriteFalse: {
		fontSize: 24,
		color: '#d3d3d3', // Gray
	},
});

export default ListItem;
