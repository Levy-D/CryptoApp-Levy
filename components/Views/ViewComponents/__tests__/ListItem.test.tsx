/* eslint-disable camelcase */
import React from 'react';
import ListItem, {styles} from '../ListItem';
import {fireEvent, mockStore, render} from '../../../../test-utils/test-utils';
import {CmcCryptoCurrency} from '../../../Interfaces/ICoinMarketCapModel';
import {
	setCryptoDataItem,
	setIsFavoriteToFalse,
	setIsFavoriteToTrue,
} from '../../../Redux/Slices/CryptoData';

let item: CmcCryptoCurrency;
let mockNavigation: any;

beforeEach(() => {
	// Arrange
	item = {
		isFavorite: false,
		id: 1,
		name: 'Bitcoin',
		symbol: 'BTC',
		slug: 'bitcoin',
		num_market_pairs: 8196,
		date_added: '2013-04-28T00:00:00.000Z',
		tags: [
			'mineable',
			'pow',
			'sha-256',
			'store-of-value',
			'state-channel',
			'coinbase-ventures-portfolio',
			'three-arrows-capital-portfolio',
			'polychain-capital-portfolio',
			'binance-labs-portfolio',
			'blockchain-capital-portfolio',
			'boostvc-portfolio',
			'cms-holdings-portfolio',
			'dcg-portfolio',
			'dragonfly-capital-portfolio',
			'electric-capital-portfolio',
			'fabric-ventures-portfolio',
			'framework-ventures-portfolio',
			'galaxy-digital-portfolio',
			'huobi-capital-portfolio',
			'alameda-research-portfolio',
			'a16z-portfolio',
			'1confirmation-portfolio',
			'winklevoss-capital-portfolio',
			'usv-portfolio',
			'placeholder-ventures-portfolio',
			'pantera-capital-portfolio',
			'multicoin-capital-portfolio',
			'paradigm-portfolio',
		],
		max_supply: 21000000,
		circulating_supply: 18905700,
		total_supply: 18905700,
		platform: null,
		cmc_rank: 1,
		last_updated: '2021-12-20T10:09:02.000Z',
		quote: {
			EUR: {
				price: 40869.528198498185,
				volume_24h: 25698098196.932026,
				volume_change_24h: 14.7957,
				percent_change_1h: -0.27651847,
				percent_change_24h: -2.7854497,
				percent_change_7d: -7.21300143,
				percent_change_30d: -21.32371954,
				percent_change_60d: -30.32862831,
				percent_change_90d: 6.69622995,
				market_cap: 772667039262.3472,
				market_cap_dominance: 40.7111,
				fully_diluted_market_cap: 858260092168.4641,
				last_updated: '2021-12-20T10:09:50.000Z',
			},
			USD: {
				price: 48734.99367730162,
				volume_24h: 29240768804.409405,
				volume_change_24h: -0.3396,
				percent_change_1h: 0.45347868,
				percent_change_24h: 4.90980381,
				percent_change_7d: 3.89633682,
				percent_change_30d: -18.53145527,
				percent_change_60d: -20.51408224,
				percent_change_90d: 12.42318288,
				market_cap: 921430381117.02,
				market_cap_dominance: 40.8726,
				fully_diluted_market_cap: 1023434867223.33,
				last_updated: '2021-12-21T16:54:02.000Z',
			},
		},
	};
	mockNavigation = {navigate: jest.fn()};
});

it('renders correctly', () => {
	// Arrange

	// Act
	const tree = render(
		<ListItem item={item} navigation={mockNavigation} />,
	).toJSON();

	// Assert
	expect(tree).toMatchSnapshot();
});

describe('Favorite Icon', () => {
	it('renders with a gray star when isFavorite = false', () => {
		// Arrange
		item.isFavorite = false;
		const style = styles.favIconIsFavoriteFalse;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
		);
		const icon = getByTestId('icon-star');

		// Assert
		expect(icon).toHaveStyle(style);
	});

	it('renders with a gold star when isFavorite = true', () => {
		// Arrange
		item.isFavorite = true;
		const style = styles.favIconIsFavoriteTrue;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
		);
		const icon = getByTestId('icon-star');

		// Assert
		expect(icon).toHaveStyle(style);
	});

	it('dispatches setIsFavoriteToTrue if isFavorite = false', () => {
		// Arrange
		item.isFavorite = false;
		const store = mockStore();
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{store},
		);
		const icon = getByTestId('icon-star');

		// Act
		fireEvent.press(icon);

		// Assert
		expect(store.dispatch).toHaveBeenCalledWith(setIsFavoriteToTrue(item));
	});

	it('dispatches setIsFavoriteToFalse if isFavorite = true', () => {
		// Arrange
		item.isFavorite = true;
		const store = mockStore();
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{store},
		);
		const icon = getByTestId('icon-star');

		// Act
		fireEvent.press(icon);

		// Assert
		expect(store.dispatch).toHaveBeenCalledWith(setIsFavoriteToFalse(item));
	});

	// it('dispatches nothing if isFavorite = undefined', () => {
	// 	// Arrange
	// 	item.isFavorite = undefined;
	// 	const store = mockStore();
	// 	const {getByTestId} = render(
	// 		<ListItem item={item} navigation={mockNavigation} />,
	// 		{store},
	// 	);
	// 	const icon = getByTestId('icon-star');

	// 	// Act
	// 	fireEvent.press(icon);

	// 	// Assert
	// 	expect(store.dispatch).toBeCalledTimes(0);
	// });
});

describe('Chainlink Highlighting', () => {
	beforeEach(() => {
		// Arrange
		item = {
			isFavorite: false,
			id: 1975,
			name: 'Chainlink',
			symbol: 'LINK',
			slug: 'chainlink',
			num_market_pairs: 458,
			date_added: '2017-09-20T00:00:00.000Z',
			tags: [
				'platform',
				'defi',
				'oracles',
				'smart-contracts',
				'substrate',
				'polkadot',
				'binance-smart-chain',
				'polkadot-ecosystem',
				'avalanche-ecosystem',
				'solana-ecosystem',
				'framework-ventures-portfolio',
				'polygon-ecosystem',
				'fantom-ecosystem',
				'cardano-ecosystem',
				'web3',
				'near-protocol-ecosystem',
			],
			max_supply: 1000000000,
			circulating_supply: 467009550.4397637,
			total_supply: 1000000000,
			platform: {
				id: 1027,
				name: 'Ethereum',
				symbol: 'ETH',
				slug: 'ethereum',
				token_address: '0x514910771af9ca656af840dff83e8264ecf986ca',
			},
			cmc_rank: 22,
			last_updated: '2021-12-21T09:28:11.000Z',
			quote: {
				USD: {
					price: 16.970193904688774,
					volume_24h: 635504559.417901,
					volume_change_24h: 7.3977,
					percent_change_1h: -0.25999981,
					percent_change_24h: 3.91910039,
					percent_change_7d: 3.74857498,
					percent_change_30d: -32.08758899,
					percent_change_60d: -34.74246881,
					percent_change_90d: -15.98782675,
					market_cap: 7925242626.304322,
					market_cap_dominance: 0.3983,
					fully_diluted_market_cap: 16970193904.691591,
					last_updated: '2021-12-21T09:29:56.000Z',
				},
			},
		};
	});

	it('should render highlighted when the ID matches that of chainlink and stateHighlightChainlink is TRUE and percent is positive', () => {
		// Arrange
		const styleListItem = styles.listItemChainlink;
		const styleListItemName = styles.listItemNameChainlink;
		const styleListItemPrice = styles.listItemPriceChainlink;
		const styleListItemPercentPositive =
			styles.listItemPercentPositiveChainlink;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{
				preloadedState: {
					userSettings: {highlightChainlink: true, displayEUR: false},
				},
			},
		);
		const touchableOpacity = getByTestId('touchable-opacity');
		const itemName = getByTestId('name');
		const itemPrice = getByTestId('price');
		const itemPercent = getByTestId('percent');

		// Assert
		expect(touchableOpacity).toHaveStyle(styleListItem);
		expect(itemName).toHaveStyle(styleListItemName);
		expect(itemPrice).toHaveStyle(styleListItemPrice);
		expect(itemPercent).toHaveStyle(styleListItemPercentPositive);
	});

	it('should render highlighted when the ID matches that of chainlink and stateHighlightChainlink is TRUE and percent is negative', () => {
		// Arrange
		item.quote.USD!.percent_change_24h = -1;
		const styleListItem = styles.listItemChainlink;
		const styleListItemName = styles.listItemNameChainlink;
		const styleListItemPrice = styles.listItemPriceChainlink;
		const styleListItemPercentNegative =
			styles.listItemPercentNegativeChainlink;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{
				preloadedState: {
					userSettings: {highlightChainlink: true, displayEUR: false},
				},
			},
		);
		const touchableOpacity = getByTestId('touchable-opacity');
		const itemName = getByTestId('name');
		const itemPrice = getByTestId('price');
		const itemPercent = getByTestId('percent');

		// Assert
		expect(touchableOpacity).toHaveStyle(styleListItem);
		expect(itemName).toHaveStyle(styleListItemName);
		expect(itemPrice).toHaveStyle(styleListItemPrice);
		expect(itemPercent).toHaveStyle(styleListItemPercentNegative);
	});

	it('should NOT render into the color ChainlinkBlue when the ID matches that of chainlink and stateHighlightChainlink is FALSE', () => {
		// Arrange
		const styleListItem = styles.listItem;
		const styleListItemName = styles.listItemName;
		const styleListItemPrice = styles.listItemPrice;
		const styleListItemPercentPositive = styles.listItemPercentPositive;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{
				preloadedState: {
					userSettings: {highlightChainlink: false, displayEUR: false},
				},
			},
		);
		const touchableOpacity = getByTestId('touchable-opacity');
		const itemName = getByTestId('name');
		const itemPrice = getByTestId('price');
		const itemPercent = getByTestId('percent');

		// Assert
		expect(touchableOpacity).toHaveStyle(styleListItem);
		expect(itemName).toHaveStyle(styleListItemName);
		expect(itemPrice).toHaveStyle(styleListItemPrice);
		expect(itemPercent).toHaveStyle(styleListItemPercentPositive);
	});

	it('should NOT render into the color ChainlinkBlue when the ID NOT matches that of chainlink and stateHighlightChainlink is FALSE', () => {
		// Arrange
		item.id = 1;
		const styleListItem = styles.listItem;
		const styleListItemName = styles.listItemName;
		const styleListItemPrice = styles.listItemPrice;
		const styleListItemPercentPositive = styles.listItemPercentPositive;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{
				preloadedState: {
					userSettings: {highlightChainlink: false, displayEUR: false},
				},
			},
		);
		const touchableOpacity = getByTestId('touchable-opacity');
		const itemName = getByTestId('name');
		const itemPrice = getByTestId('price');
		const itemPercent = getByTestId('percent');

		// Assert
		expect(touchableOpacity).toHaveStyle(styleListItem);
		expect(itemName).toHaveStyle(styleListItemName);
		expect(itemPrice).toHaveStyle(styleListItemPrice);
		expect(itemPercent).toHaveStyle(styleListItemPercentPositive);
	});

	it('should NOT render into the color ChainlinkBlue when the ID NOT matches that of chainlink and stateHighlightChainlink is TRUE', () => {
		// Arrange
		item.id = 1;
		const styleListItem = styles.listItem;
		const styleListItemName = styles.listItemName;
		const styleListItemPrice = styles.listItemPrice;
		const styleListItemPercentPositive = styles.listItemPercentPositive;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{
				preloadedState: {
					userSettings: {highlightChainlink: true, displayEUR: false},
				},
			},
		);
		const touchableOpacity = getByTestId('touchable-opacity');
		const itemName = getByTestId('name');
		const itemPrice = getByTestId('price');
		const itemPercent = getByTestId('percent');

		// Assert
		expect(touchableOpacity).toHaveStyle(styleListItem);
		expect(itemName).toHaveStyle(styleListItemName);
		expect(itemPrice).toHaveStyle(styleListItemPrice);
		expect(itemPercent).toHaveStyle(styleListItemPercentPositive);
	});
});

describe('Valuta', () => {
	it('should render with EUR values', () => {
		// Arrange
		item.quote.EUR!.price = 2000;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{
				preloadedState: {
					userSettings: {highlightChainlink: false, displayEUR: true},
				},
			},
		);
		const itemPrice = getByTestId('price');

		// Assert
		expect(itemPrice).toHaveTextContent('€2000.00');
	});

	it('should render with USD values', () => {
		// Arrange
		item.quote.USD!.price = 1000;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{
				preloadedState: {
					userSettings: {highlightChainlink: false, displayEUR: false},
				},
			},
		);
		const itemPrice = getByTestId('price');

		// Assert
		expect(itemPrice).toHaveTextContent('$1000.00');
	});

	it('should round down to 2 digits after the comma if the price is 1', () => {
		// Arrange
		item.quote.USD!.price = 1;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
		);
		const itemPrice = getByTestId('price');

		// Assert
		expect(itemPrice).toHaveTextContent('$1.00');
	});

	it('should round down to 2 digits after the comma if the price is higher than 1', () => {
		// Arrange
		item.quote.USD!.price = 1.01;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
		);
		const itemPrice = getByTestId('price');

		// Assert
		expect(itemPrice).toHaveTextContent('$1.01');
	});

	it('should show 4 digits to precision if the price is lower than 1', () => {
		// Arrange
		item.quote.USD!.price = 0.995456464;

		// Act
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
		);
		const itemPrice = getByTestId('price');

		// Assert
		expect(itemPrice).toHaveTextContent('$0.9955');
	});
});

describe('Press cryptocurrency', () => {
	it('dispatches setCryptoDataItem', () => {
		// Arrange
		const store = mockStore();
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
			{store},
		);
		const crypto = getByTestId('touchable-opacity');

		// Act
		fireEvent.press(crypto);

		// Assert
		expect(store.dispatch).toHaveBeenCalledWith(setCryptoDataItem(item));
	});

	it('navigates to Crypto Details', () => {
		// Arrange
		const {getByTestId} = render(
			<ListItem item={item} navigation={mockNavigation} />,
		);
		const crypto = getByTestId('touchable-opacity');

		// Act
		fireEvent.press(crypto);

		// Assert
		expect(mockNavigation.navigate).toHaveBeenCalledWith('Crypto Details');
	});
});
