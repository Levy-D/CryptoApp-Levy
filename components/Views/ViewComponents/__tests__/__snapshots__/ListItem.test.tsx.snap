// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#f8f8f8",
      "borderBottomWidth": 1,
      "borderColor": "#eee",
      "fontSize": 16,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 10,
    }
  }
  testID="touchable-opacity"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "width": 130,
        }
      }
      testID="name"
    >
      Bitcoin
    </Text>
    <Text
      style={
        Object {
          "width": 100,
        }
      }
      testID="price"
    >
      $48734.99
    </Text>
    <Text
      style={
        Object {
          "color": "green",
          "textAlign": "right",
          "width": 75,
        }
      }
      testID="percent"
    >
      4.91
      %
    </Text>
    <Text
      allowFontScaling={false}
      onPress={[Function]}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "#d3d3d3",
            "fontSize": 24,
          },
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
      testID="icon-star"
    >
      ï€…
    </Text>
  </View>
</View>
`;
